File: web-sdks/dart/lib/src/core/exceptions.dart
/// Comprehensive exception system for GroupVAN SDK
///
/// This file defines a hierarchy of exceptions that provide detailed error
/// information and enable proper error handling strategies.
library exceptions;

/// Base exception for all GroupVAN SDK errors
abstract class GroupVanException implements Exception {
  /// Human-readable error message
  final String message;

  /// Optional additional context about the error
  final Map<String, dynamic>? context;

  /// Timestamp when the error occurred
  final DateTime timestamp;

  GroupVanException(this.message, {this.context, DateTime? timestamp})
    : timestamp = timestamp ?? DateTime.now();

  @override
  String toString() => 'GroupVanException: $message';
}

/// Network-related exceptions
class NetworkException extends GroupVanException {
  /// The endpoint that was being accessed
  final String? endpoint;

  /// The HTTP method used
  final String? method;

  NetworkException(
    super.message, {
    this.endpoint,
    this.method,
    super.context,
    super.timestamp,
  });

  @override
  String toString() =>
      'NetworkException: $message${endpoint != null ? ' (endpoint: $endpoint)' : ''}';
}

/// HTTP response exceptions
class HttpException extends GroupVanException {
  /// HTTP status code
  final int statusCode;

  /// The endpoint that was being accessed
  final String endpoint;

  /// The HTTP method used
  final String method;

  /// Response body if available
  final String? responseBody;

  HttpException(
    super.message, {
    required this.statusCode,
    required this.endpoint,
    required this.method,
    this.responseBody,
    super.context,
    super.timestamp,
  });

  /// Whether this is a client error (4xx)
  bool get isClientError => statusCode >= 400 && statusCode < 500;

  /// Whether this is a server error (5xx)
  bool get isServerError => statusCode >= 500;

  /// Whether this error is retryable
  bool get isRetryable =>
      isServerError || statusCode == 429 || statusCode == 408;

  @override
  String toString() =>
      'HttpException: $message (Status: $statusCode, Method: $method, Endpoint: $endpoint)';
}

/// Authentication and authorization exceptions
class AuthenticationException extends GroupVanException {
  /// Type of authentication error
  final AuthErrorType errorType;

  AuthenticationException(
    super.message, {
    required this.errorType,
    super.context,
    super.timestamp,
  });

  @override
  String toString() => 'AuthenticationException: $message (Type: $errorType)';
}

/// Types of authentication errors
enum AuthErrorType {
  invalidToken,
  expiredToken,
  missingToken,
  insufficientPermissions,
  invalidCredentials,
  accountNotLinked,
}

/// Validation exceptions for user input
class ValidationException extends GroupVanException {
  /// List of validation errors
  final List<ValidationError> errors;

  ValidationException(
    super.message, {
    required this.errors,
    super.context,
    super.timestamp,
  });

  /// Get all error messages
  List<String> get errorMessages => errors.map((e) => e.message).toList();

  @override
  String toString() =>
      'ValidationException: $message (${errors.length} errors)';
}

/// Individual validation error
class ValidationError {
  /// Field name that failed validation
  final String field;

  /// Error message
  final String message;

  /// The invalid value
  final dynamic value;

  /// Validation rule that failed
  final String? rule;

  const ValidationError({
    required this.field,
    required this.message,
    this.value,
    this.rule,
  });

  @override
  String toString() => '$field: $message';
}

/// Configuration exceptions
class ConfigurationException extends GroupVanException {
  /// Configuration key that caused the error
  final String? configKey;

  ConfigurationException(
    super.message, {
    this.configKey,
    super.context,
    super.timestamp,
  });

  @override
  String toString() =>
      'ConfigurationException: $message${configKey != null ? ' (key: $configKey)' : ''}';
}

/// Rate limiting exceptions
class RateLimitException extends GroupVanException {
  /// Number of seconds until retry is allowed
  final int? retryAfterSeconds;

  /// Current rate limit window
  final Duration? window;

  /// Number of requests allowed in window
  final int? limit;

  RateLimitException(
    super.message, {
    this.retryAfterSeconds,
    this.window,
    this.limit,
    super.context,
    super.timestamp,
  });

  @override
  String toString() =>
      'RateLimitException: $message${retryAfterSeconds != null ? ' (retry after: ${retryAfterSeconds}s)' : ''}';
}

/// Data parsing exceptions
class DataException extends GroupVanException {
  /// Type of data that failed to parse
  final String? dataType;

  /// Original data that caused the error
  final dynamic originalData;

  DataException(
    super.message, {
    this.dataType,
    this.originalData,
    super.context,
    super.timestamp,
  });

  @override
  String toString() =>
      'DataException: $message${dataType != null ? ' (type: $dataType)' : ''}';
}

File: web-sdks/javascript/src/core/exceptions.js
/**
 * Comprehensive exception system for GroupVAN SDK
 *
 * This file defines a hierarchy of exceptions that provide detailed error
 * information and enable proper error handling strategies.
 */

/**
 * Base exception for all GroupVAN SDK errors
 */
export class GroupVanException extends Error {
  /**
   * @param {string} message - Human-readable error message
   * @param {Object} [options] - Additional options
   * @param {Object} [options.context] - Optional additional context about the error
   * @param {Date} [options.timestamp] - Timestamp when the error occurred
   */
  constructor(message, { context = null, timestamp = null } = {}) {
    super(message);
    this.name = 'GroupVanException';
    this.context = context;
    this.timestamp = timestamp || new Date();
  }

  toString() {
    return `GroupVanException: ${this.message}`;
  }
}

/**
 * Network-related exceptions
 */
export class NetworkException extends GroupVanException {
  /**
   * @param {string} message - Error message
   * @param {Object} [options] - Additional options
   * @param {string} [options.endpoint] - The endpoint that was being accessed
   * @param {string} [options.method] - The HTTP method used
   * @param {Object} [options.context] - Optional additional context
   * @param {Date} [options.timestamp] - Timestamp when the error occurred
   */
  constructor(message, { endpoint = null, method = null, context = null, timestamp = null } = {}) {
    super(message, { context, timestamp });
    this.name = 'NetworkException';
    this.endpoint = endpoint;
    this.method = method;
  }

  toString() {
    return `NetworkException: ${this.message}${this.endpoint ? ` (endpoint: ${this.endpoint})` : ''}`;
  }
}

/**
 * HTTP response exceptions
 */
export class HttpException extends GroupVanException {
  /**
   * @param {string} message - Error message
   * @param {Object} options - Required options
   * @param {number} options.statusCode - HTTP status code
   * @param {string} options.endpoint - The endpoint that was being accessed
   * @param {string} options.method - The HTTP method used
   * @param {string} [options.responseBody] - Response body if available
   * @param {Object} [options.context] - Optional additional context
   * @param {Date} [options.timestamp] - Timestamp when the error occurred
   */
  constructor(message, { statusCode, endpoint, method, responseBody = null, context = null, timestamp = null }) {
    super(message, { context, timestamp });
    this.name = 'HttpException';
    this.statusCode = statusCode;
    this.endpoint = endpoint;
    this.method = method;
    this.responseBody = responseBody;
  }

  /**
   * Whether this is a client error (4xx)
   * @returns {boolean}
   */
  get isClientError() {
    return this.statusCode >= 400 && this.statusCode < 500;
  }

  /**
   * Whether this is a server error (5xx)
   * @returns {boolean}
   */
  get isServerError() {
    return this.statusCode >= 500;
  }

  /**
   * Whether this error is retryable
   * @returns {boolean}
   */
  get isRetryable() {
    return this.isServerError || this.statusCode === 429 || this.statusCode === 408;
  }

  toString() {
    return `HttpException: ${this.message} (Status: ${this.statusCode}, Method: ${this.method}, Endpoint: ${this.endpoint})`;
  }
}

/**
 * Types of authentication errors
 * @enum {string}
 */
export const AuthErrorType = {
  INVALID_TOKEN: 'invalidToken',
  EXPIRED_TOKEN: 'expiredToken',
  MISSING_TOKEN: 'missingToken',
  INSUFFICIENT_PERMISSIONS: 'insufficientPermissions',
  INVALID_CREDENTIALS: 'invalidCredentials',
  ACCOUNT_NOT_LINKED: 'accountNotLinked',
};

/**
 * Authentication and authorization exceptions
 */
export class AuthenticationException extends GroupVanException {
  /**
   * @param {string} message - Error message
   * @param {Object} options - Required options
   * @param {string} options.errorType - Type of authentication error from AuthErrorType
   * @param {Object} [options.context] - Optional additional context
   * @param {Date} [options.timestamp] - Timestamp when the error occurred
   */
  constructor(message, { errorType, context = null, timestamp = null }) {
    super(message, { context, timestamp });
    this.name = 'AuthenticationException';
    this.errorType = errorType;
  }

  toString() {
    return `AuthenticationException: ${this.message} (Type: ${this.errorType})`;
  }
}

/**
 * Individual validation error
 */
export class ValidationError {
  /**
   * @param {Object} options
   * @param {string} options.field - Field name that failed validation
   * @param {string} options.message - Error message
   * @param {*} [options.value] - The invalid value
   * @param {string} [options.rule] - Validation rule that failed
   */
  constructor({ field, message, value = null, rule = null }) {
    this.field = field;
    this.message = message;
    this.value = value;
    this.rule = rule;
  }

  toString() {
    return `${this.field}: ${this.message}`;
  }
}

/**
 * Validation exceptions for user input
 */
export class ValidationException extends GroupVanException {
  /**
   * @param {string} message - Error message
   * @param {Object} options - Required options
   * @param {ValidationError[]} options.errors - List of validation errors
   * @param {Object} [options.context] - Optional additional context
   * @param {Date} [options.timestamp] - Timestamp when the error occurred
   */
  constructor(message, { errors, context = null, timestamp = null }) {
    super(message, { context, timestamp });
    this.name = 'ValidationException';
    this.errors = errors;
  }

  /**
   * Get all error messages
   * @returns {string[]}
   */
  get errorMessages() {
    return this.errors.map(e => e.message);
  }

  toString() {
    return `ValidationException: ${this.message} (${this.errors.length} errors)`;
  }
}

/**
 * Configuration exceptions
 */
export class ConfigurationException extends GroupVanException {
  /**
   * @param {string} message - Error message
   * @param {Object} [options] - Additional options
   * @param {string} [options.configKey] - Configuration key that caused the error
   * @param {Object} [options.context] - Optional additional context
   * @param {Date} [options.timestamp] - Timestamp when the error occurred
   */
  constructor(message, { configKey = null, context = null, timestamp = null } = {}) {
    super(message, { context, timestamp });
    this.name = 'ConfigurationException';
    this.configKey = configKey;
  }

  toString() {
    return `ConfigurationException: ${this.message}${this.configKey ? ` (key: ${this.configKey})` : ''}`;
  }
}

/**
 * Rate limiting exceptions
 */
export class RateLimitException extends GroupVanException {
  /**
   * @param {string} message - Error message
   * @param {Object} [options] - Additional options
   * @param {number} [options.retryAfterSeconds] - Number of seconds until retry is allowed
   * @param {number} [options.windowMs] - Current rate limit window in milliseconds
   * @param {number} [options.limit] - Number of requests allowed in window
   * @param {Object} [options.context] - Optional additional context
   * @param {Date} [options.timestamp] - Timestamp when the error occurred
   */
  constructor(message, { retryAfterSeconds = null, windowMs = null, limit = null, context = null, timestamp = null } = {}) {
    super(message, { context, timestamp });
    this.name = 'RateLimitException';
    this.retryAfterSeconds = retryAfterSeconds;
    this.windowMs = windowMs;
    this.limit = limit;
  }

  toString() {
    return `RateLimitException: ${this.message}${this.retryAfterSeconds ? ` (retry after: ${this.retryAfterSeconds}s)` : ''}`;
  }
}

/**
 * Data parsing exceptions
 */
export class DataException extends GroupVanException {
  /**
   * @param {string} message - Error message
   * @param {Object} [options] - Additional options
   * @param {string} [options.dataType] - Type of data that failed to parse
   * @param {*} [options.originalData] - Original data that caused the error
   * @param {Object} [options.context] - Optional additional context
   * @param {Date} [options.timestamp] - Timestamp when the error occurred
   */
  constructor(message, { dataType = null, originalData = null, context = null, timestamp = null } = {}) {
    super(message, { context, timestamp });
    this.name = 'DataException';
    this.dataType = dataType;
    this.originalData = originalData;
  }

  toString() {
    return `DataException: ${this.message}${this.dataType ? ` (type: ${this.dataType})` : ''}`;
  }
}
